cli/src/stake.rs

```diff
@@ -2488,7 +2491,8 @@ pub fn make_cli_reward(
 ) -> Option<CliEpochReward> {
     let wallclock_epoch_duration = epoch_end_time.checked_sub(epoch_start_time)?;
     if reward.post_balance > reward.amount {
-        let rate_change = reward.amount as f64 / (reward.post_balance - reward.amount) as f64;
+        let rate_change =
+            reward.amount as f64 / (reward.post_balance.saturating_sub(reward.amount)) as f64;
```

Comment:
Since it was just checked that `post_balance > amount`, can you do checked variants and unwrap? Can you also do `checked_div` instead of `/`?

Reply:
I thought that checked math is worse than saturating or wrapping, as it requires more work.  And an `unwrap()` is a potential panic, which also needs to be optimized away.
Why do you prefer it to the saturating math?

As `post_balance` is strictly greater than the `amount`, `post_balance - amount` is non-zero.
So there seems to be no risk in dividing by zero and no need to verify the division anymore?
Also, `checked_div` is only for integers.  `f64` will produce an `inf` value if the divisor is zero.

Reply:
That's a good point... we could just do wrapping in that case. And TIL about floats not doing checked math!

Reply:
Why wrapping?
Saturating is as fast and seems a bit safer to me %)
Wrapping is the least reasonable choice of all, from the mathematical standpoint, is it not? %)

Reply:
I ran a little test modifying the BPF program at https://github.com/solana-labs/solana-program-library/tree/master/examples/rust/transfer-lamports.

You can get the compute units by running `cargo test-sbf`, so I modified the addition and subtraction to do saturating, checked + unwrap, wrapping, and the raw operation. Here were the compute units consumed:

* saturating: 516
* checked + unwrap: 509
* wrapping / raw: 502

This is a simple example, but surprisingly, saturating had the worst performance! From what I understand, since wrapping has no additional code to check or bound the result, it should always be fastest, which is why I proposed it. Where did you see that saturating is as fast as wrapping?

Reply:
BPF is a pretty special target.  It has just a few native instructions, so most logic needs to be expressed explicitly.  Not sure if this is a good test.

On x86 you get this: https://gcc.godbolt.org/z/n74Eq3WsT
I thought that there is a native instruction on x86 for saturated addition, but it is actually expressed via 3 instructions.
I still think that it would be executed almost at a speed very close to the wrapped 
addition.
Though, looking at the checked math assembly, I think it would probably be pretty quick as well.

Reply:
I figured, it is interesting to measure the execution times with [`criterion`](https://bheisler.github.io/criterion.rs/book/criterion_rs.html).

Here is the test:
```rust
use criterion::{criterion_group, criterion_main, Criterion};
use rand::Rng;
use std::hint::black_box;

#[inline(never)]
fn add_wrapping(v1: u64, v2: u64) -> u64 {
    v1 + v2
}

#[inline(never)]
fn add_saturating(v1: u64, v2: u64) -> u64 {
    v1.saturating_add(v2)
}

#[inline(never)]
fn add_checked(v1: u64, v2: u64) -> u64 {
    v1.checked_add(v2).unwrap()
}

fn bench_addition(c: &mut Criterion) {
    let mut group = c.benchmark_group("Addition");
    let mut rng = rand::thread_rng();

    let (v1, v2): (u64, u64) = (
        rng.gen_range(0..u64::MAX / 2),
        rng.gen_range(0..u64::MAX / 2),
    );
    group.bench_with_input("wrapping", &(v1, v2), |b, (v1, v2)| {
        b.iter(|| add_wrapping(black_box(*v1), black_box(*v2)))
    });
    group.bench_with_input("saturating", &(v1, v2), |b, (v1, v2)| {
        b.iter(|| add_saturating(black_box(*v1), black_box(*v2)))
    });
    group.bench_with_input("checked", &(v1, v2), |b, (v1, v2)| {
        b.iter(|| add_checked(black_box(*v1), black_box(*v2)))
    });

    group.finish();
}

criterion_group!(benches, bench_addition);
criterion_main!(benches);
```

Here are the timings:

```
❯ cargo bench
   Compiling rust v0.1.0 (/home/ilya/works/tests/solana/rust)
     Running benches/addition.rs (target/release/deps/addition-d2e120044f9f187c)
Gnuplot not found, using plotters backend
Addition/wrapping       time:   [1.0921 ns 1.0958 ns 1.1011 ns]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
Addition/saturating     time:   [1.0713 ns 1.0736 ns 1.0764 ns]
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  9 (9.00%) high severe
Addition/checked        time:   [1.0676 ns 1.0681 ns 1.0686 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe
```

If I remove `#[inline(never)]` I get this:
```
❯ cargo bench
   Compiling rust v0.1.0 (/home/ilya/works/tests/solana/rust)
    Finished bench [optimized] target(s) in 0.69s
     Running benches/addition.rs (target/release/deps/addition-d2e120044f9f187c)
Gnuplot not found, using plotters backend
Addition/wrapping       time:   [644.34 ps 646.44 ps 648.66 ps]
                        change: [-40.978% -40.773% -40.599%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 22 outliers among 100 measurements (22.00%)
  18 (18.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
Addition/saturating     time:   [652.29 ps 653.95 ps 655.57 ps]
                        change: [-39.387% -39.235% -39.085%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
Addition/checked        time:   [636.37 ps 636.80 ps 637.22 ps]
                        change: [-40.448% -40.369% -40.274%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe

```

(It actually failed a few times in the `checked` case, as a random choice of
`v1` and `v2` did cause an overflow sometimes, before I added `gen_range()`.).


So, I would say, there are no practical differences in performance.

---

If you want to play with this benchmark yourself, put the above into
`benches/addition.rs`, create `Cargo.toml` as shown below, and you can measure
by running `cargo bench`.

```Cargo.toml
[package]
name = "my-benchmark"
version = "0.1.0"
edition = "2021"

[dependencies]
criterion = "0.5.1"
rand = "0.8.5"

[dev-dependencies]

[[bench]]
name = "addition"
harness = false
```

Reply:
Thanks for explaining everything and providing the test cases! I ran them also on my machine out of curiosity, and got roughly similar results. Here are the second test results, with inlining:

```
Benchmarking Addition/wrapping: Collecting 100 samples in estimated 5.0000 s (9.8B iterati
Addition/wrapping       time:   [509.83 ps 510.01 ps 510.30 ps]
                        change: [-32.013% -31.483% -30.977%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  1 (1.00%) high mild
  5 (5.00%) high severe
Benchmarking Addition/saturating: Collecting 100 samples in estimated 5.0000 s (7.0B itera
Addition/saturating     time:   [643.98 ps 660.47 ps 676.12 ps]
                        change: [-15.419% -13.328% -11.514%] (p = 0.00 < 0.05)
                        Performance has improved.
Benchmarking Addition/checked: Collecting 100 samples in estimated 5.0000 s (9.2B iteratio
Addition/checked        time:   [539.47 ps 540.05 ps 540.78 ps]
                        change: [-29.139% -28.566% -28.020%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe
```

Thanks to the previous check in the code, we can do the wrapping math safely, which is why I suggested it. Since it's not performance-sensitive, we can go with whatever's most readable. `saturating` is a totally safe option, so we can do that.

I have a slight preference for Brooks's suggestion for checked / expect to give the reason why it's safe, but I'll leave it up to you.

Reply:
I think the `if` condition is informative enough here.
I agree, in general, it better to err on the side of the code being more verbose.  It then captures more of the author intent.
But there is also a negative side to it.
The more verbose the code is, the harder it is to actually see the original intent.
It gets "noisier" and thus can be harder to read.
So, while for important logic, `checked/expect` is certainly the best, I am not sure if it should be the always used.

On a general note, I am not trying to improve all the existing mathematical operations in the `cli` module.
I am just trying to re-enable the clippy checks.
Forcing people who are going to modify code, and add new code, to think a bit about their math and choose the right function.

Reevaluating every single existing choice is a lot of work.
And it is not really my intent.
My take is, wrapping math failure, which is what the code would do currently, is either very similar or strictly worse, compared to the failure you get in the saturating math case.
So, I if I re-enable the warnings and use saturating math in most places, it keeps the code quality either the same or slightly better.  But is strictly better for future modifications.

A few locations that I do read carefully enough, I can probably improve.
But I did not analyze every case.

I'm explaining it, as I have a bunch more changes in the queue, that modify at least a few dozens more math operations in the `cli`.
I should have explained my intent better in the PR description when asking for reviews, sorry about that.

I will certainly accept any comments to the subsequent PRs, but just keep in mind the original intent, when you are reviewing them.